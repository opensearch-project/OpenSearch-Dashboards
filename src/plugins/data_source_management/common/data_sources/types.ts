/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { ToastInputFields, ErrorToastOptions } from 'src/core/public/notifications';
// eslint-disable-next-line @osd/eslint/no-restricted-paths
import type { SavedObject } from 'src/core/server';

export interface IDataSource {
  title: string;
  id?: string;
  type?: string;
  endpoint: string;
}

export interface IDataSourceAttributes {
  title: string;
  id: string;
  type: string;
  endpoint: string;
}

export type OnNotification = (toastInputFields: ToastInputFields) => void;
export type OnError = (error: Error, toastInputFields: ErrorToastOptions) => void;

export interface UiSettingsCommon {
  get: (key: string) => Promise<any>;
  getAll: () => Promise<Record<string, any>>;
  set: (keu: string, value: any) => Promise<void>;
  remove: (key: string) => Promise<void>;
}

export interface SavedObjectsClientCommonFindArgs {
  type: string | string[];
  perPage?: number;
  // fields?: string[];
  // search?: string;
  // searchFields?: string[];
}

export interface SavedObjectsClientCommon {
  find: <T = unknown>(options: SavedObjectsClientCommonFindArgs) => Promise<Array<SavedObject<T>>>;
  get: <T = unknown>(type: string, id: string) => Promise<SavedObject<T>>;
  update: <T = unknown>(
    type: string,
    id: string,
    attributes: Record<string, any>,
    options: Record<string, any>
  ) => Promise<SavedObject<T>>;
  create: (
    type: string,
    attributes: Record<string, any>,
    options: Record<string, any>
  ) => Promise<SavedObject>;
  delete: (type: string, id: string) => Promise<{}>;
}

// export interface IDataSourcesApiClient {

// }

export interface DataSourceSpec {
  id?: string;
  version?: string;
  title?: string;
  type?: string;
  endpoint?: string;
}

export type { SavedObject };
