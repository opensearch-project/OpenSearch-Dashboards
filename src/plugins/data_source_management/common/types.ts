/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export enum ManageableBy {
  All = 'all',
  DashboardAdmin = 'dashboard_admin',
  None = 'none',
}
