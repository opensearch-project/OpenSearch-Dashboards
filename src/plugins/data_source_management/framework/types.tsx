/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiComboBoxOptionOption } from '@elastic/eui';

export enum DirectQueryLoadingStatus {
  SUCCESS = 'success',
  FAILED = 'failed',
  RUNNING = 'running',
  SCHEDULED = 'scheduled',
  CANCELED = 'canceled',
  WAITING = 'waiting',
  INITIAL = 'initial',
}

export interface DirectQueryRequest {
  query: string;
  lang: string;
  datasource: string;
  sessionId?: string;
}

export type AccelerationStatus = 'ACTIVE' | 'INACTIVE';

export interface PermissionsConfigurationProps {
  roles: Role[];
  selectedRoles: Role[];
  setSelectedRoles: React.Dispatch<React.SetStateAction<Role[]>>;
  layout: 'horizontal' | 'vertical';
  hasSecurityAccess: boolean;
}

export interface TableColumn {
  name: string;
  dataType: string;
}

export interface Acceleration {
  name: string;
  status: AccelerationStatus;
  type: string;
  database: string;
  table: string;
  destination: string;
  dateCreated: number;
  dateUpdated: number;
  index: string;
  sql: string;
}

export interface AssociatedObject {
  tableName: string;
  datasource: string;
  id: string;
  name: string;
  database: string;
  type: AssociatedObjectIndexType;
  accelerations: CachedAcceleration[] | AssociatedObject;
  columns?: CachedColumn[];
}

export type Role = EuiComboBoxOptionOption;

export type DatasourceType = 'S3GLUE' | 'PROMETHEUS';

export interface S3GlueProperties {
  'glue.indexstore.opensearch.uri': string;
  'glue.indexstore.opensearch.region': string;
}

export interface PrometheusProperties {
  'prometheus.uri': string;
}

export type DatasourceStatus = 'ACTIVE' | 'DISABLED';

export interface DatasourceDetails {
  allowedRoles: string[];
  name: string;
  connector: DatasourceType;
  description: string;
  properties: S3GlueProperties | PrometheusProperties;
  status: DatasourceStatus;
}

interface AsyncApiDataResponse {
  status: string;
  schema?: Array<{ name: string; type: string }>;
  datarows?: any;
  total?: number;
  size?: number;
  error?: string;
}

export interface AsyncApiResponse {
  data: {
    ok: boolean;
    resp: AsyncApiDataResponse;
  };
}

export type PollingCallback = (statusObj: AsyncApiResponse) => void;

export type AssociatedObjectIndexType = AccelerationIndexType | 'table';

export type AccelerationIndexType = 'skipping' | 'covering' | 'materialized';

export type LoadCacheType = 'databases' | 'tables' | 'accelerations' | 'tableColumns';

export enum CachedDataSourceStatus {
  Updated = 'Updated',
  Failed = 'Failed',
  Empty = 'Empty',
}

export interface CachedColumn {
  fieldName: string;
  dataType: string;
}

export interface CachedTable {
  name: string;
  columns?: CachedColumn[];
}

export interface CachedDatabase {
  name: string;
  tables: CachedTable[];
  lastUpdated: string; // date string in UTC format
  status: CachedDataSourceStatus;
}

export interface CachedDataSource {
  name: string;
  lastUpdated: string; // date string in UTC format
  status: CachedDataSourceStatus;
  databases: CachedDatabase[];
  dataSourceMDSId?: string;
}

export interface DataSourceCacheData {
  version: string;
  dataSources: CachedDataSource[];
}

export interface CachedAcceleration {
  flintIndexName: string;
  type: AccelerationIndexType;
  database: string;
  table: string;
  indexName: string;
  autoRefresh: boolean;
  status: string;
}

export interface CachedAccelerationByDataSource {
  name: string;
  accelerations: CachedAcceleration[];
  lastUpdated: string; // date string in UTC format
  status: CachedDataSourceStatus;
  dataSourceMDSId?: string;
}

export interface AccelerationsCacheData {
  version: string;
  dataSources: CachedAccelerationByDataSource[];
}

export interface PollingSuccessResult {
  schema: Array<{ name: string; type: string }>;
  datarows: Array<Array<string | number | boolean>>;
}

export type AsyncPollingResult = PollingSuccessResult | null;

export type AggregationFunctionType = 'count' | 'sum' | 'avg' | 'max' | 'min' | 'window.start';

export interface MaterializedViewColumn {
  id: string;
  functionName: AggregationFunctionType;
  functionParam?: string;
  fieldAlias?: string;
}

export type SkippingIndexAccMethodType = 'PARTITION' | 'VALUE_SET' | 'MIN_MAX' | 'BLOOM_FILTER';

export interface SkippingIndexRowType {
  id: string;
  fieldName: string;
  dataType: string;
  accelerationMethod: SkippingIndexAccMethodType;
}

export interface DataTableFieldsType {
  id: string;
  fieldName: string;
  dataType: string;
}

export interface RefreshIntervalType {
  refreshWindow: number;
  refreshInterval: string;
}

export interface WatermarkDelayType {
  delayWindow: number;
  delayInterval: string;
}

export interface GroupByTumbleType {
  timeField: string;
  tumbleWindow: number;
  tumbleInterval: string;
}

export interface MaterializedViewQueryType {
  columnsValues: MaterializedViewColumn[];
  groupByTumbleValue: GroupByTumbleType;
}

export interface FormErrorsType {
  dataSourceError: string[];
  databaseError: string[];
  dataTableError: string[];
  skippingIndexError: string[];
  coveringIndexError: string[];
  materializedViewError: string[];
  indexNameError: string[];
  primaryShardsError: string[];
  replicaShardsError: string[];
  refreshIntervalError: string[];
  checkpointLocationError: string[];
  watermarkDelayError: string[];
}

export type AccelerationRefreshType = 'autoInterval' | 'manual' | 'manualIncrement';

export interface CreateAccelerationForm {
  dataSource: string;
  database: string;
  dataTable: string;
  dataTableFields: DataTableFieldsType[];
  accelerationIndexType: AccelerationIndexType;
  skippingIndexQueryData: SkippingIndexRowType[];
  coveringIndexQueryData: string[];
  materializedViewQueryData: MaterializedViewQueryType;
  accelerationIndexName: string;
  primaryShardsCount: number;
  replicaShardsCount: number;
  refreshType: AccelerationRefreshType;
  checkpointLocation: string | undefined;
  watermarkDelay: WatermarkDelayType;
  refreshIntervalOptions: RefreshIntervalType;
  formErrors: FormErrorsType;
}

export interface LoadCachehookOutput {
  loadStatus: DirectQueryLoadingStatus;
  startLoading: (params: StartLoadingParams) => void;
  stopLoading: () => void;
}

export interface StartLoadingParams {
  dataSourceName: string;
  dataSourceMDSId?: string;
  databaseName?: string;
  tableName?: string;
}

export interface RenderAccelerationFlyoutParams {
  dataSourceName: string;
  dataSourceMDSId?: string;
  databaseName?: string;
  tableName?: string;
  handleRefresh?: () => void;
}

export interface RenderAssociatedObjectsDetailsFlyoutParams {
  tableDetail: AssociatedObject;
  dataSourceName: string;
  handleRefresh?: () => void;
  dataSourceMDSId?: string;
}

export interface RenderAccelerationDetailsFlyoutParams {
  acceleration: CachedAcceleration;
  dataSourceName: string;
  handleRefresh?: () => void;
  dataSourceMDSId?: string;
}

// Integration types

export interface StaticAsset {
  annotation?: string;
  path: string;
}

export interface IntegrationWorkflow {
  name: string;
  label: string;
  description: string;
  enabled_by_default: boolean;
}

export interface IntegrationStatics {
  logo?: StaticAsset;
  gallery?: StaticAsset[];
  darkModeLogo?: StaticAsset;
  darkModeGallery?: StaticAsset[];
}

export interface IntegrationComponent {
  name: string;
  version: string;
}

export type SupportedAssetType = 'savedObjectBundle' | 'query';

export interface IntegrationAsset {
  name: string;
  version: string;
  extension: string;
  type: SupportedAssetType;
  workflows?: string[];
}

export interface IntegrationConfig {
  name: string;
  version: string;
  displayName?: string;
  license: string;
  type: string;
  labels?: string[];
  author?: string;
  description?: string;
  sourceUrl?: string;
  workflows?: IntegrationWorkflow[];
  statics?: IntegrationStatics;
  components: IntegrationComponent[];
  assets: IntegrationAsset[];
  sampleData?: {
    path: string;
  };
}

export interface AvailableIntegrationsList {
  hits: IntegrationConfig[];
}

export interface AssetReference {
  assetType: string;
  assetId: string;
  isDefaultAsset: boolean;
  description: string;
  status?: string;
}

export interface IntegrationInstanceResult extends IntegrationInstance {
  id: string;
  status: string;
}

export interface IntegrationInstance {
  name: string;
  templateName: string;
  dataSource: string;
  creationDate: string;
  assets: AssetReference[];
}

export interface AvailableIntegrationsList {
  hits: IntegrationConfig[];
}

export interface IntegrationInstancesSearchResult {
  hits: IntegrationInstanceResult[];
}

export type ParsedIntegrationAsset =
  | { type: 'savedObjectBundle'; workflows?: string[]; data: object[] }
  | { type: 'query'; workflows?: string[]; query: string; language: string };

export type Result<T, E = Error> =
  | { ok: true; value: T; error?: undefined }
  | { ok: false; error: E; value?: undefined };
