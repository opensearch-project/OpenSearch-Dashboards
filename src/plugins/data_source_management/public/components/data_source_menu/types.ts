/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  MountPoint,
  NotificationsStart,
  SavedObjectsClientContract,
  SavedObject,
  IUiSettingsClient,
} from '../../../../../core/public';
import { DataSourceAttributes } from '../../types';

export interface DataSourceOption {
  id: string;
  label?: string;
}

export interface DataSourceGroupLabelOption extends DataSourceOption {
  label: string;
  isGroupLabel: true;
}

export interface DataSourceBaseConfig {
  fullWidth: boolean;
  disabled?: boolean;
}

export interface DataSourceMenuProps<T = any> {
  componentType: DataSourceComponentType;
  componentConfig: T;
  hideLocalCluster?: boolean;
  uiSettings?: IUiSettingsClient;
  setMenuMountPoint?: (menuMount: MountPoint | undefined) => void;
}

export const DataSourceComponentType = {
  DataSourceSelectable: 'DataSourceSelectable',
  DataSourceView: 'DataSourceView',
  DataSourceAggregatedView: 'DataSourceAggregatedView',
  DataSourceMultiSelectable: 'DataSourceMultiSelectable',
} as const;

export type DataSourceComponentType = typeof DataSourceComponentType[keyof typeof DataSourceComponentType];

export interface DataSourceViewConfig extends DataSourceBaseConfig {
  activeOption: DataSourceOption[];
  savedObjects?: SavedObjectsClientContract;
  notifications?: NotificationsStart;
}

export interface DataSourceAggregatedViewConfig extends DataSourceBaseConfig {
  savedObjects: SavedObjectsClientContract;
  notifications: NotificationsStart;
  activeDataSourceIds?: string[];
  displayAllCompatibleDataSources?: boolean;
  dataSourceFilter?: (dataSource: SavedObject<DataSourceAttributes>) => boolean;
}

export interface DataSourceSelectableConfig extends DataSourceBaseConfig {
  onSelectedDataSources: (dataSources: DataSourceOption[]) => void;
  savedObjects: SavedObjectsClientContract;
  notifications: NotificationsStart;
  activeOption?: DataSourceOption[];
  dataSourceFilter?: (dataSource: SavedObject<DataSourceAttributes>) => boolean;
}

export interface DataSourceMultiSelectableConfig extends DataSourceBaseConfig {
  onSelectedDataSources: (dataSources: DataSourceOption[]) => void;
  savedObjects: SavedObjectsClientContract;
  notifications: NotificationsStart;
}
