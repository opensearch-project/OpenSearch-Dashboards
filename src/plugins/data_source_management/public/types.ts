/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  ChromeStart,
  ApplicationStart,
  IUiSettingsClient,
  OverlayStart,
  SavedObjectsStart,
  NotificationsStart,
  DocLinksStart,
  HttpSetup,
} from 'src/core/public';
import { ManagementAppMountParams } from 'src/plugins/management/public';
import { SavedObjectAttributes } from 'src/core/types';
import { i18n } from '@osd/i18n';
import { SigV4ServiceName } from '../../data_source/common/data_sources';
import { OpenSearchDashboardsReactContextValue } from '../../opensearch_dashboards_react/public';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface DataSourceManagementPluginStart {}

export interface DataSourceManagementContext {
  chrome: ChromeStart;
  application: ApplicationStart;
  savedObjects: SavedObjectsStart;
  uiSettings: IUiSettingsClient;
  notifications: NotificationsStart;
  overlays: OverlayStart;
  http: HttpSetup;
  docLinks: DocLinksStart;
  setBreadcrumbs: ManagementAppMountParams['setBreadcrumbs'];
}

export interface DataSourceTableItem {
  id: string;
  title: string;
  description: string;
  sort: string;
}

export interface ToastMessageItem {
  id: string;
  defaultMessage: string;
  success?: boolean;
}

export type DataSourceManagementContextValue = OpenSearchDashboardsReactContextValue<
  DataSourceManagementContext
>;

/* Datasource types */
export enum AuthType {
  NoAuth = 'no_auth',
  UsernamePasswordType = 'username_password',
  SigV4 = 'sigv4',
}

export const credentialSourceOptions = [
  {
    value: AuthType.NoAuth,
    text: i18n.translate('dataSourceManagement.credentialSourceOptions.NoAuthentication', {
      defaultMessage: 'No authentication',
    }),
  },
  {
    value: AuthType.UsernamePasswordType,
    text: i18n.translate('dataSourceManagement.credentialSourceOptions.UsernamePassword', {
      defaultMessage: 'Username & Password',
    }),
  },
  {
    value: AuthType.SigV4,
    text: i18n.translate('dataSourceManagement.credentialSourceOptions.AwsSigV4', {
      defaultMessage: 'AWS SigV4',
    }),
  },
];

export const sigV4ServiceOptions = [
  {
    value: SigV4ServiceName.OpenSearch,
    text: i18n.translate('dataSourceManagement.SigV4ServiceOptions.OpenSearch', {
      defaultMessage: 'Amazon OpenSearch Service',
    }),
  },
  {
    value: SigV4ServiceName.OpenSearchServerless,
    text: i18n.translate('dataSourceManagement.SigV4ServiceOptions.OpenSearchServerless', {
      defaultMessage: 'Amazon OpenSearch Serverless',
    }),
  },
];

export interface DataSourceAttributes extends SavedObjectAttributes {
  title: string;
  description?: string;
  endpoint?: string;
  auth: {
    type: AuthType;
    credentials: UsernamePasswordTypedContent | SigV4Content | undefined;
  };
}

export interface UsernamePasswordTypedContent extends SavedObjectAttributes {
  username: string;
  password?: string;
}

export interface SigV4Content extends SavedObjectAttributes {
  accessKey: string;
  secretKey: string;
  region: string;
  service?: SigV4ServiceName;
}
