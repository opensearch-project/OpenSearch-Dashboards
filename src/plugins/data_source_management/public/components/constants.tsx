/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { i18n } from '@osd/i18n';
import { DataSourceOption } from './data_source_menu/types';

export const LocalCluster: DataSourceOption = {
  label: i18n.translate('dataSource.localCluster', {
    defaultMessage: 'Local cluster',
  }),
  id: '',
};
