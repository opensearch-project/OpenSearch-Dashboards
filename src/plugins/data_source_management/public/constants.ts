/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { DirectQueryDatasourceType } from './types';

export const QUERY_RESTRICTED = 'query-restricted';
export const QUERY_ALL = 'query-all';

export const DatasourceTypeToDisplayName: { [key in DirectQueryDatasourceType]: string } = {
  PROMETHEUS: 'Prometheus',
  S3GLUE: 'Amazon S3',
};

export const PROMETHEUS_URL = 'Prometheus';
export const AMAZON_S3_URL = 'AmazonS3AWSGlue';
export const OPENSEARCH_URL = 'OpenSearch';

export const UrlToDatasourceType: { [key: string]: DirectQueryDatasourceType } = {
  [PROMETHEUS_URL]: 'PROMETHEUS',
  [AMAZON_S3_URL]: 'S3GLUE',
};
