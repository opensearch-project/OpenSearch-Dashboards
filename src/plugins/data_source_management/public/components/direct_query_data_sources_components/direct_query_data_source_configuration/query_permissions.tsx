/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiComboBox,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormRow,
  EuiRadioGroup,
  EuiSpacer,
  EuiText,
} from '@elastic/eui';
import React, { useState } from 'react';
import { QUERY_ALL, QUERY_RESTRICTED } from '../../../constants';
import { PermissionsConfigurationProps } from '../../../types';

export const QueryPermissionsConfiguration = (props: PermissionsConfigurationProps) => {
  const { roles, selectedRoles, setSelectedRoles, layout, hasSecurityAccess } = props;

  const [selectedAccessLevel, setSelectedAccessLevel] = useState(
    selectedRoles.length ? QUERY_RESTRICTED : QUERY_ALL
  );

  const accessLevelOptions = [
    {
      id: QUERY_RESTRICTED,
      label: 'Restricted - accessible by users with specific OpenSearch roles',
      disabled: !hasSecurityAccess,
    },
    {
      id: QUERY_ALL,
      label: 'Admin only - only accessible by the admin',
      disabled: !hasSecurityAccess,
    },
  ];

  const ConfigureRoles = () => {
    return (
      <div>
        <EuiSpacer size="s" />
        <EuiText>OpenSearch Roles</EuiText>
        <EuiText size="xs">
          Select one or more OpenSearch roles that can query this data connection.
        </EuiText>
        <EuiFormRow
          isInvalid={selectedRoles.length === 0}
          error={
            selectedRoles.length === 0
              ? 'Select an OpenSearch role or roles that will have query access to this data source'
              : undefined
          }
        >
          <EuiComboBox
            placeholder="Select one or more options"
            options={roles}
            selectedOptions={selectedRoles}
            onChange={setSelectedRoles}
            isClearable={true}
            data-test-subj="query-permissions-combo-box"
            isInvalid={selectedRoles.length === 0}
          />
        </EuiFormRow>
      </div>
    );
  };

  return (
    <EuiFlexItem>
      <EuiFlexGroup direction={layout === 'horizontal' ? 'row' : 'column'}>
        <EuiFlexItem>
          <EuiText>
            <h3>Query permissions</h3>
          </EuiText>
          <EuiText size="s">
            <p>
              Control which OpenSearch roles have permission to query and index data from this data
              source.
            </p>
          </EuiText>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiRadioGroup
            options={accessLevelOptions}
            idSelected={selectedAccessLevel}
            onChange={(id) => {
              if (id === QUERY_ALL) {
                setSelectedRoles([]);
              }
              setSelectedAccessLevel(id);
            }}
            name="query-radio-group"
            legend={{
              children: <span>Query access level</span>,
            }}
          />
          {selectedAccessLevel === QUERY_RESTRICTED && <ConfigureRoles />}
        </EuiFlexItem>
      </EuiFlexGroup>
    </EuiFlexItem>
  );
};
