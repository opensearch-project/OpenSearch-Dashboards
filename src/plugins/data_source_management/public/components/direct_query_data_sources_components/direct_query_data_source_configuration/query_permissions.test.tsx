/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState } from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { EuiComboBox, EuiRadioGroup } from '@elastic/eui';
import { QUERY_ALL, QUERY_RESTRICTED } from '../../../constants';
import { QueryPermissionsConfiguration } from './query_permissions';
import { Role } from '../../../types';

const roles: Role[] = [
  { label: 'Admin', value: 'admin' },
  { label: 'User', value: 'user' },
];

describe('QueryPermissionsConfiguration', () => {
  let wrapper: ReactWrapper;

  beforeEach(() => {
    const TestComponent = () => {
      const [selectedRoles, setSelectedRoles] = useState<Role[]>([]);
      return (
        <QueryPermissionsConfiguration
          roles={roles}
          selectedRoles={selectedRoles}
          setSelectedRoles={setSelectedRoles}
          layout="vertical"
          hasSecurityAccess={true}
        />
      );
    };

    wrapper = mount(<TestComponent />);
  });

  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('renders the radio group with correct options', () => {
    const radioGroup = wrapper.find(EuiRadioGroup);
    expect(radioGroup.exists()).toBe(true);
    expect(radioGroup.prop('options')).toHaveLength(2);
  });

  test('changes access level state on radio group change', async () => {
    const radioGroup = wrapper.find(EuiRadioGroup);
    expect(radioGroup.prop('idSelected')).toBe(QUERY_ALL);

    await act(async () => {
      radioGroup.prop('onChange')!(QUERY_RESTRICTED);
    });

    wrapper.update();
    expect(wrapper.find(EuiRadioGroup).prop('idSelected')).toBe(QUERY_RESTRICTED);
  });

  test('renders EuiComboBox when QUERY_RESTRICTED is selected', async () => {
    await act(async () => {
      wrapper.find(EuiRadioGroup).prop('onChange')!(QUERY_RESTRICTED);
    });

    wrapper.update();
    expect(wrapper.find(EuiComboBox).exists()).toBe(true);
  });

  test('validates EuiComboBox correctly', async () => {
    await act(async () => {
      wrapper.find(EuiRadioGroup).prop('onChange')!(QUERY_RESTRICTED);
    });

    wrapper.update();
    const comboBox = wrapper.find(EuiComboBox);

    expect(comboBox.prop('isInvalid')).toBe(true);

    await act(async () => {
      comboBox.prop('onChange')!([{ label: 'Admin', value: 'admin' }]);
    });

    wrapper.update();

    const updatedComboBox = wrapper.find(EuiComboBox);
    expect(updatedComboBox.prop('isInvalid')).toBe(false);
  });
});
