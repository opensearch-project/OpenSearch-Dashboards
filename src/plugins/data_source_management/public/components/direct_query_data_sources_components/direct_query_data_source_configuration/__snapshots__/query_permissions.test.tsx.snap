// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryPermissionsConfiguration renders correctly 1`] = `
<TestComponent>
  <QueryPermissionsConfiguration
    hasSecurityAccess={true}
    layout="vertical"
    roles={
      Array [
        Object {
          "label": "Admin",
          "value": "admin",
        },
        Object {
          "label": "User",
          "value": "user",
        },
      ]
    }
    selectedRoles={Array []}
    setSelectedRoles={[Function]}
  >
    <EuiFlexItem>
      <div
        className="euiFlexItem"
      >
        <EuiFlexGroup
          direction="column"
        >
          <div
            className="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
          >
            <EuiFlexItem>
              <div
                className="euiFlexItem"
              >
                <EuiText>
                  <div
                    className="euiText euiText--medium"
                  >
                    <h3>
                      Query permissions
                    </h3>
                  </div>
                </EuiText>
                <EuiText
                  size="s"
                >
                  <div
                    className="euiText euiText--small"
                  >
                    <p>
                      Control which OpenSearch roles have permission to query and index data from this data source.
                    </p>
                  </div>
                </EuiText>
              </div>
            </EuiFlexItem>
            <EuiFlexItem>
              <div
                className="euiFlexItem"
              >
                <EuiRadioGroup
                  idSelected="query-all"
                  legend={
                    Object {
                      "children": <span>
                        Query access level
                      </span>,
                      "compressed": undefined,
                    }
                  }
                  name="query-radio-group"
                  onChange={[Function]}
                  options={
                    Array [
                      Object {
                        "disabled": false,
                        "id": "query-restricted",
                        "label": "Restricted - accessible by users with specific OpenSearch roles",
                      },
                      Object {
                        "disabled": false,
                        "id": "query-all",
                        "label": "Admin only - only accessible by the admin",
                      },
                    ]
                  }
                >
                  <EuiFormFieldset
                    legend={
                      Object {
                        "children": <span>
                          Query access level
                        </span>,
                        "compressed": undefined,
                      }
                    }
                  >
                    <fieldset>
                      <EuiFormLegend>
                        <legend
                          className="euiFormLegend"
                        >
                          <span>
                            Query access level
                          </span>
                        </legend>
                      </EuiFormLegend>
                      <EuiRadio
                        checked={false}
                        className="euiRadioGroup__item"
                        disabled={false}
                        id="query-restricted"
                        key="0"
                        label="Restricted - accessible by users with specific OpenSearch roles"
                        name="query-radio-group"
                        onChange={[Function]}
                      >
                        <div
                          className="euiRadio euiRadioGroup__item"
                        >
                          <input
                            checked={false}
                            className="euiRadio__input"
                            disabled={false}
                            id="query-restricted"
                            name="query-radio-group"
                            onChange={[Function]}
                            type="radio"
                          />
                          <div
                            className="euiRadio__circle"
                          />
                          <label
                            className="euiRadio__label"
                            htmlFor="query-restricted"
                          >
                            Restricted - accessible by users with specific OpenSearch roles
                          </label>
                        </div>
                      </EuiRadio>
                      <EuiRadio
                        checked={true}
                        className="euiRadioGroup__item"
                        disabled={false}
                        id="query-all"
                        key="1"
                        label="Admin only - only accessible by the admin"
                        name="query-radio-group"
                        onChange={[Function]}
                      >
                        <div
                          className="euiRadio euiRadioGroup__item"
                        >
                          <input
                            checked={true}
                            className="euiRadio__input"
                            disabled={false}
                            id="query-all"
                            name="query-radio-group"
                            onChange={[Function]}
                            type="radio"
                          />
                          <div
                            className="euiRadio__circle"
                          />
                          <label
                            className="euiRadio__label"
                            htmlFor="query-all"
                          >
                            Admin only - only accessible by the admin
                          </label>
                        </div>
                      </EuiRadio>
                    </fieldset>
                  </EuiFormFieldset>
                </EuiRadioGroup>
              </div>
            </EuiFlexItem>
          </div>
        </EuiFlexGroup>
      </div>
    </EuiFlexItem>
  </QueryPermissionsConfiguration>
</TestComponent>
`;
